EIPS/pairings.md

```diff
@@ -0,0 +1,138 @@
+
+Pairing functions can be used to perform a limited form of multiplicatively homomorphic operations, which are necessary for current zkSNARKs. This precompile can be used to run such computations within the block gas limit. This precompiled contract only specifies a certain check, and not an evaluation of a pairing function. The reason is that the codomain of a pairing function is a rather complex field which could provide encoding problems and all known uses of pairing function in zkSNARKs only require the specified check.
+
+## Specification
+
+Add a precompiled contracts for a bilinear function on groups on the elliptic curve "alt_bn128". We will define the precompiled contract in terms of a discrete logarithm. The discrete logarithm is of course assumed to be hard to compute, but we will give an equivalent specification that makes use of elliptic curve pairing functions which can be efficiently computed below.
+
+Address: 0x8
+
+For a cyclic group `G` (written additively) of prime order `q` let `log_P: G -> F_q` be the discrete logarithm on this group with respect to a generator `P`, i.e. `log_P(x)` is the smallest non-negative integer `n` such that `n * P = x`.
+
+The precompiled contract is defined as follows, where the two groups `G_1` and `G_2` and their generators `P_1` and `P_2` are defined below. Both generators have the same prime order `q` and the actual choice of the generators does not matter, as long as they have order `q`.
+
+```
+Input: (a1, b1, a2, b2, ..., ak, bk) from (G_1 x G_2)^k
+Output: If the length of the input is incorrect or any of the inputs are not elements of
+        the respective group or are not encoded correctly, the call fails.
+        Otherwise, return one if
+        log_P1(a1) * log_P2(b1) + ... + log_P1(ak) * log_P2(bk) = 0
+        (in F_q) and zero else.
+```
+
+Note that `k` is determined from the length of the input. Following the section on the encoding below,
+`k` is the length of the input divided by `192`. If the input length is not a multiple of `192`,
+the call fails. Empty input is valid and results in returning one.
+
+In order to check that an input is an element of `G_1`, verifying the encoding of the coordinates and checking that they satisfy the curve equation (or is the encoding of infinity) is sufficient. For `G_2`, in addition to that, the order of the element has to be checked to be equal to the group order `q = 21888242871839275222246405745257275088548364400416034343698204186575808495617`.
+
+### Definition of the groups
+
+The groups `G_1` and `G_2` are cyclic groups of prime order `q = 21888242871839275222246405745257275088548364400416034343698204186575808495617` on the elliptic curve `alt_bn128` defined by the curve equation
+`Y^2 = X^3 + 3`.
+
+The group `G_1` is a cyclic group on the above curve over the field `F_p` with `p = 21888242871839275222246405745257275088696311157297823662689037894645226208583` with generator `P1 = (1, 2)`.
+
```

Comment:
Rather than specifying the groups using giant magic numbers, might be worth specifying them in terms of the BN parameter from which they are derived. See e.g. https://github.com/cdetrio/py_ecc/blob/4d7096a313b7cbff0f0103c548160c8098b4b0f4/py_ecc/optimized_bn128/parameters.py#L4-L8

Reply:
In that case, we need to specify how to identify a subgroup given BN parameters.

Reply:
A giant magic number obviously specifies a unique mathematical object.  A more indirect description must be accompanied by a proof that there is a unique thing that satisfies the indirect description.

Reply:
The BN parameter `4965661367192848881` can be used to derive the field modulus: `field_modulus = 36*(BN_PARAM_U**4) + 36*(BN_PARAM_U**3) + 24*(BN_PARAM_U**2) + 6*BN_PARAM_U + 1`.  I suggested including it because its a neat formula that took me a while to track down, but perhaps other readers don't care.

Reply:
@cdetrio please describe how that number uniquely identifies a subgroup on the curve.  Is that a shorter way to identify the group?